File: static/r/add_tif_palette.R
# From: http://rstudio-pubs-static.s3.amazonaws.com/3334_35452e4a611a4fa39ccebd26dffe4c9e.html

makePalette <- function(colourvector) {
    cmat = cbind(t(col2rgb(colourvector)), 255)
    res = apply(cmat, 1, function(x) {
        sprintf("<Entry c1=\"%s\" c2=\"%s\" c3=\"%s\" c4=\"%s\"/>", x[1], x[2],
            x[3], x[4])
    })
    res = paste(res, collapse = "\n")
    res
}

makePaletteVRT <- function(raster, colourvector) {
    s = sprintf("<VRTDataset rasterXSize=\"%s\" rasterYSize=\"%s\">\n<VRTRasterBand dataType=\"Byte\" band=\"1\">\n<ColorInterp>Palette</ColorInterp>\n<ColorTable>\n",
        ncol(raster), nrow(raster))
    p = makePalette(colourvector)
    s = paste0(s, p, "\n</ColorTable>\n</VRTRasterBand>\n</VRTDataset>\n")
    s
}

writePaletteVRT <- function(out, raster, colourvector) {
    s = makePaletteVRT(raster, colourvector)
    cat(s, file = out)
}

# writePaletteVRT("test.vrt", r, rainbow(6))

File: static/py/add_tif_palette.py
#! /usr/bin/env python
#
# Take a raster and add a palette
#
# From: http://rstudio-pubs-static.s3.amazonaws.com/3334_35452e4a611a4fa39ccebd26dffe4c9e.html
#

#  1. write an integer raster from R like this: writeRaster(r,"rint.tiff",datatype="INT1U")
#
#  2. write a palette file from R like this:  writePaletteVRT("test.vrt",r,rainbow(20))
#
#  3. run this script like this:
#      addPalette test.vrt rint.tiff
#     and it will modify rint.tiff, adding the palette.

try:
    from osgeo import gdal
except ImportError:
    import gdal

#import gdal2
import sys

def Usage():
    print('Usage: add_tif_palette.py palette_file raster_file')
    sys.exit(1)

gdal.AllRegister()
argv = gdal.GeneralCmdLineProcessor(sys.argv)
if argv is None:
    Usage()

if len(argv)!=3:
    Usage()

pal_file = argv[1]
raster_file = argv[2]

src_ds = gdal.Open(raster_file)

pal_ds = gdal.Open(pal_file)
ct = pal_ds.GetRasterBand(1).GetRasterColorTable().Clone()
src_ds.GetRasterBand(1).SetRasterColorTable(ct)

gtiff_driver = gdal.GetDriverByName('GTiff')
gtiff_driver.CreateCopy(raster_file, src_ds)
